File: src/main/java/io/scriptor/http/result/HTTPResultString.java
package io.scriptor.http.result;

import org.jetbrains.annotations.NotNull;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.util.Map;

public class HTTPResultString extends HTTPResult<String> {

    public HTTPResultString(final int statusCode) {
        super(statusCode);
    }

    public HTTPResultString(final int statusCode, final @NotNull String statusText) {
        super(statusCode, statusText);
    }

    public HTTPResultString(final int statusCode, final @NotNull String statusText, final @NotNull String value) {
        super(statusCode, statusText, value);
    }

    public HTTPResultString(final int statusCode, final @NotNull Map<String, String> headers) {
        super(statusCode, headers);
    }

    public HTTPResultString(
            final int statusCode,
            final @NotNull String statusText,
            final @NotNull Map<String, String> headers
    ) {
        super(statusCode, statusText, headers);
    }

    public HTTPResultString(
            final int statusCode,
            final @NotNull String statusText,
            final @NotNull Map<String, String> headers,
            final @NotNull String value
    ) {
        super(statusCode, statusText, headers, value);
    }

    @Override
    public int getSize() {
        return getBody().getBytes().length;
    }

    @Override
    public @NotNull InputStream getStream() {
        return new ByteArrayInputStream(getBody().getBytes());
    }
}

File: server/src/main/kotlin/dev/scriptor/server/http/result/HTTPResultString.kt
package dev.scriptor.server.http.result

import dev.scriptor.server.http.ParameterList
import java.nio.channels.Channels
import java.nio.channels.ReadableByteChannel

class HTTPResultString : HTTPResult<String> {

    private companion object {

        data class CountChannel(val count: Long, val channel: ReadableByteChannel?)

        fun create(value: String?): CountChannel {
            val count: Long
            val channel: ReadableByteChannel?
            if (value != null) {
                val buf = value.encodeToByteArray()

                count = buf.size.toLong()
                channel = Channels.newChannel(buf.inputStream())
            } else {
                count = 0L
                channel = null
            }
            return CountChannel(count, channel)
        }
    }

    private constructor(
        statusCode: Int,
        statusText: String,
        headers: ParameterList,
        countChannel: CountChannel,
    ) : super(statusCode, statusText, headers, 0L, countChannel.count, countChannel.channel)

    constructor(
        statusCode: Int = 200,
        statusText: String = "OK",
        headers: ParameterList = ParameterList(),
        value: String? = null,
    ) : this(statusCode, statusText, headers, create(value))
}
